BJS2-77856

SAM19991 opened the pull request:
Условия задачи

В user_service должна быть возможность загрузить данные из csv-файла и распарсить их в java-объект. Например, университет хочет зарегистрировать своих студентов на нашей платформе. Они присылают нам список в формате .csv с данными студентов. Наша задача — получить данные из csv в java-объектах, преобразовать их во внутренний формат и сохранить всех студентов в базе. Тестовые файлы (csv и json-схема) находятся в папке resources. JSON-схема предоставлена для ознакомления — на её основе вам предстоит вручную написать нужные POJO-классы.

Тех. детали

В UserController написать метод, который принимает объект MultipartFile для csv. В контроллере нужно получить данные из файла, например, в InputStream и передать их в метод UserService. Можно использовать CsvMapper.

По json-схеме нужно вручную создать pojo-классы. Используйте JSON-схему как подсказку для полей и структуры. Класс будет содержать необходимые поля для преобразования данных из csv в java-объекты (например, Person с полями для имени, почты и других данных).  

Когда Pojo готовы, нужно преобразовать строки, полученные из csv в java-объекты, используя эти pojo. 1 строка из файла = 1 объект.

Затем используем Mapstruct, чтобы преобразовать Person в User. Подсказка по преобразованию ниже.

Добавить автогенерацию паролей. В будущем каждый пользователь сможет изменить пароль под себя.

Если страна, в которой живёт пользователь, уже есть в базе, присвоить её пользователю, если не — создать новую страну.

После маппинга каждый объект сохранить в таблицу Users. Теперь пользователи зарегистрированы!

Подумайте, можно ли использовать многопоточность для решения этой задачи.

Для всех классов с логикой должны быть написаны unit-тесты, покрывающие все их публичные методы. Используем JUnit и Mockito.

Подробности использования готовых компонентов

Если вы не знакомы с логикой работы с БД (пройдем позже на буткемпе) и не готовы разбираться с ней самостоятельно прямо сейчас, то в проекте уже заранее создан интерфейс UserRepository и CountryRepository со всеми методами, которые могут вам пригодиться в процессе реализации этого функционала.

Вот эти методы:

save(User user) — сохраняет пользователя в базу.

save(Country country) — сохраняет страну в базу.

findAll() — возвращает список всех стран.

SAM19991 committed:
BJS2-77856
Парсим файлы пользователей

SAM19991 committed:
Merge branch 'triton-master-stream10' into develop/BJS2-77856

kashkademan commented on src/main/java/school/faang/user_service/controller/UserController.java:
```diff
@@ -24,8 +27,13 @@ public UserDto getUser(@PathVariable Long userId) {
     }
 
-    @PostMapping
+    @PostMapping("/by-ids")
```
мне кажется тут можно пустой путь оставить типа хотим гет всех пользователей а передать параметром только лист.

SAM19991 replied:
это уже не мой код, тут Danila писал ))

kashkademan commented on src/main/java/school/faang/user_service/controller/UserController.java:
```diff
@@ -24,8 +27,13 @@ public UserDto getUser(@PathVariable Long userId) {
         return userService.getUsersByIds(ids);
     }
+
+    @PostMapping("/upload-csv")
```
может просто оставить upload типа можно же грузить например еще хоть с помощь чего угодно, а дальше уже (в будущем) можно будет проверять чего там загрузилось. например эксель может быть или еще что то.

kashkademan commented on src/main/java/school/faang/user_service/dto/csv/StudentCsvDto.java:
```diff
@@ -0,0 +1,94 @@
+@NoArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class StudentCsvDto {
+    @JsonProperty("firstName")
+    @NotBlank(message = "First name cannot be blank")
+    private String firstName;
+    @JsonProperty("lastName")
```
как то все слипается, может перенос строки после переменной?))

SAM19991 replied:
готово

kashkademan commented on src/main/java/school/faang/user_service/service/user_service/UserServiceImpl.java:
```diff
@@ -0,0 +1,128 @@
+            List<User> processedUsers = new ArrayList<>();
+            List<String> errors = new ArrayList<>();
+            for (StudentCsvDto studentDto : students) {
+                try {
```
А зачем еще раз try внутри try? или так надо? функционально оно если вылетит исключение то уже все равно помается.

SAM19991 replied:
idea ругается , требует чтобы было так. Внутренний try-catch нужен для resilient processing - продолжаем обработку других студентов, даже если один не обработался

kashkademan commented on src/main/java/school/faang/user_service/service/user_service/UserServiceImpl.java:
```diff
@@ -0,0 +1,128 @@
+
+    private List<StudentCsvDto> readCsvFile(InputStream inputStream) {
+        CsvSchema schema = csvMapper.schemaFor(StudentCsvDto.class)
+                .withColumnSeparator(',')
+                .withLineSeparator("\n")
```
может это в константы куда то если они могут меняться эти значения?

kashkademan commented on src/main/java/school/faang/user_service/service/user_service/UserServiceImpl.java:
```diff
@@ -0,0 +1,128 @@
+                        (existingCountry, newCountry) -> existingCountry));
+    }
+
+    private Country getOrCreateCountry(String countryName, Map<String, Country> countriesMap) {
```
что то не очень найди или создай. может разделить. попробуй получить, и если нету отдельной функцией создать?

SAM19991 replied:
Готово

kashkademan reviewed (changes requested):
Привет!) написал пару комментов.

SAM19991 committed:
исправлены все замечания

miked4king commented on src/main/java/school/faang/user_service/controller/UserController.java:
```diff
@@ -6,26 +6,37 @@
     }
 
-    @PostMapping
+    @PostMapping("/by-ids")
```
/by-ids - давай уберем

SAM19991 replied:
Но ведь это писал Данила, я просто добавил   @PostMapping("/upload") в этот класс

miked4king commented on src/main/java/school/faang/user_service/service/user_service_upload/UserServiceUploadImpl.java:
```diff
@@ -0,0 +1,139 @@
+
+    @Override
+    public CsvUploadResponseDto processStudentsCsv(MultipartFile file) {
+        try {
```
нам не нужно весь код заворачивать в try catch)) давай уберем отсюда
у тебя скорее всего требует обработки file.getInputStream().

будет хорошо немного изменить метод readCsvFile, передавать в него MultipartFile 
и использовать trywithresources, примерно так 
CsvSchema schema = csvMapper.schemaFor(StudentCsvDto.class).withHeader();
        MappingIterator<StudentCsvDto> iterator;
        try (InputStream inputStream = file.getInputStream()) {
            iterator = csvMapper
                    .readerFor(StudentCsvDto.class)
                    .with(schema)
                    .readValues(inputStream);
            return iterator.readAll();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

SAM19991 replied:
готово

miked4king commented on src/main/java/school/faang/user_service/service/user_service_upload/UserServiceUploadImpl.java:
```diff
@@ -0,0 +1,139 @@
+            Map<String, Country> countriesMap = getAllCountriesMap();
+            List<User> processedUsers = new ArrayList<>();
+            List<String> errors = new ArrayList<>();
+            for (StudentCsvDto studentDto : students) {
+                try {
+                    User user = processUser(studentDto, countriesMap);
+                    processedUsers.add(user);
+                } catch (Exception e) {
+                    errors.add("Error processing student " + studentDto.getEmail() + ": " + e.getMessage());
+                    log.error("Error processing student  {}", studentDto.getEmail(), e);
+
+                }
```
неплохая задумка обработать все ошибки из csv файла.
может вместо email что то другое передавать? какой нибудь id

miked4king commented on src/main/java/school/faang/user_service/service/user_service_upload/UserServiceUploadImpl.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    private List<StudentCsvDto> readCsvFile(InputStream inputStream) {
+        CsvSchema schema = csvMapper.schemaFor(StudentCsvDto.class).withHeader();
```
как переработать метод написал выше

miked4king commented on src/main/java/school/faang/user_service/service/user_service_upload/PasswordGenerator.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.security.SecureRandom;
+
+@Component
```
предлагаю сделать его утилитным классом, вынесем в отдельный пакет в корне util
над классом повесим аннотацию @UtilityClass и будем вызывать generatePassword через имя класса

miked4king commented on src/main/java/school/faang/user_service/service/user_service_upload/PasswordGenerator.java:
```diff
@@ -0,0 +1,20 @@
+package school.faang.user_service.service.user_service_upload;
```
странное название пакета user_service_upload
общепринято называть одним словом (существительным)

miked4king commented on src/main/java/school/faang/user_service/service/user_service_upload/UserServiceUploadImpl.java:
```diff
@@ -0,0 +1,139 @@
+            response.setProcessedCount(processedUsers.size());
+            response.setErrors(errors);
+            response.setErrorCount(errors.size());
+            return response;
```
комментарий по методу:
1) исходя из названий вообще не понятно что он делает)
2) если мы собираем ошибки List<String> errors  то думаю будет логично бросить исключение и завершить работу. Либо все студенты успешно сохранятся либо ни один

miked4king commented on src/main/java/school/faang/user_service/service/user_service_upload/UserServiceUploadImpl.java:
```diff
@@ -0,0 +1,139 @@
+@Service
+@Slf4j
+@RequiredArgsConstructor
+@Transactional
```
может на конкретном методе стоит поставить?

SAM19991 replied:
готово

miked4king commented on src/main/java/school/faang/user_service/mapper/CsvUserMapper.java:
```diff
@@ -0,0 +1,25 @@
+import school.faang.user_service.entity.User;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface CsvUserMapper {
```
может в master уже есть UserMapper?
если есть то тогда лучше в нем добавить методы, чем создавать новый класс

SAM19991 committed:
refactor: improve upload service structure and CSV processing
- Relocate PasswordGenerator to util package
- Enhance CSV processing with better validation and error handling
- Merge CsvUserMapper functionality into UserMapper
- Renamed user_service_upload to upload
